### PR TITLE
[SPARK-18533] Raise correct error upon specification of schema for datasource tables created using CTAS

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
Fixes the inconsistency of error raised between data source and hive serde 
tables when schema is specified in CTAS scenario. In the process the grammar for 
create table (datasource) is simplified.

**before:**
``` SQL
spark-sql> create table t2 (c1 int, c2 int) using parquet as select * from t1;
Error in query: 
mismatched input 'as' expecting {<EOF>, '.', 'OPTIONS', 'CLUSTERED', 'PARTITIONED'}(line 1, pos 64)

== SQL ==
create table t2 (c1 int, c2 int) using parquet as select * from t1
----------------------------------------------------------------^^^
```

**After:**
```SQL
spark-sql> create table t2 (c1 int, c2 int) using parquet as select * from t1
         > ;
Error in query: 
Operation not allowed: Schema may not be specified in a Create Table As Select (CTAS) statement(line 1, pos 0)

== SQL ==
create table t2 (c1 int, c2 int) using parquet as select * from t1
^^^
```
## How was this patch tested?
Added a new test in CreateTableAsSelectSuite

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/sources/CreateTableAsSelectSuite.scala [hunk 1]
```diff
@@ -249,4 +249,13 @@ class CreateTableAsSelectSuite
       }
     }
   }
+
+  test("specifying the column list for CTAS") {
+    withTable("t") {
+      val e = intercept[AnalysisException] {
+          sql("CREATE TABLE t (a int, b int) USING parquet AS SELECT 1, 2")
+        }.getMessage
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/CreateTableAsSelectSuite.scala]: Nit: Indents.

**Comment by gatorsmile** [comment]
Please fix the PR description.

**Comment by gatorsmile** [comment]
LGTM pending test. 

cc @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/sources/CreateTableAsSelectSuite.scala (hunk 1 above)
Comment (dilipbiswal) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/CreateTableAsSelectSuite.scala]: @gatorsmile Thank you. Fixed ..

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 2]
```diff
@@ -1052,7 +1077,8 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
             "CTAS statement."
           operationNotAllowed(errorMessage, ctx)
         }
-        // Just use whatever is projected in the select statement as our schema
```
Comment (tejasapatil) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: why not keep this ?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/CreateTableAsSelectSuite.scala [hunk 3]
```diff
@@ -249,4 +249,13 @@ class CreateTableAsSelectSuite
       }
     }
   }
+
+  test("specifying the column list for CTAS") {
+    withTable("t") {
+      val e = intercept[AnalysisException] {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/CreateTableAsSelectSuite.scala]: `ParserException`?

**Comment by cloud-fan** [comment]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/sources/CreateTableAsSelectSuite.scala (hunk 3 above)
Comment (dilipbiswal) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/CreateTableAsSelectSuite.scala]: @cloud-fan Thanks !! I have made the change.

**Comment by dilipbiswal** [comment]
retest this please

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 4]
```diff
@@ -71,11 +71,7 @@ statement
     | createTableHeader ('(' colTypeList ')')? tableProvider
         (OPTIONS tablePropertyList)?
         (PARTITIONED BY partitionColumnNames=identifierList)?
-        bucketSpec?                                                    #createTableUsing
-    | createTableHeader tableProvider
-        (OPTIONS tablePropertyList)?
-        (PARTITIONED BY partitionColumnNames=identifierList)?
-        bucketSpec? AS? query                                          #createTableUsing
+        bucketSpec? (AS? query)?                                       #createTableUsing
```
Comment (viirya) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: AS is optional?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 5]
```diff
@@ -322,7 +322,26 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
   }
 
   /**
-   * Create a [[CreateTable]] logical plan.
+   * Create a table, returning a [[CreateTable]] logical plan.
+   *
+   * This is used to create data source tables, which is handled through
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: nit: I know this is copied from `visitCreateTable`'s comment. So the sentence "which is ..." looks weird and redundant here.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 4 above)
Comment (dilipbiswal) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: @viirya yeah. AS is optional and (AS query) is optional.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 5 above)
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: @viirya Thanks. I have fixed the comment. Let me know if it looks ok to you now.

**Comment by gatorsmile** [comment]
```
== SQL ==
explain extended create table t2 (c1 int, c2 int) using parquet as select * from t1
----------------------------------------------------------------^^^
```

The PR description is still not right. This does not match the query you issued. Could you update it?

**Comment by dilipbiswal** [comment]
@gatorsmile Thanks .. updated !!

**Comment by gatorsmile** [comment]
Thanks! Merging to master/2.1!

**Pull request closed without merging**